### PR TITLE
Positive jacobian

### PR DESCRIPTION
**Pull request by Huangzizhou**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -268,6 +272,10 @@ target_link_libraries(polyfem PUBLIC Glob::Glob)
 include(units)
 target_link_libraries(polyfem PUBLIC units::units)
 
+# GMP
```
Comment (teseoch) [inline CMakeLists.txt:275]: this is a bit scary, can we have it optional?

File: src/polyfem/solver/SolveData.cpp [hunk 2]
```diff
@@ -34,11 +34,13 @@ namespace polyfem::solver
 
 		// Elastic form
 		const int n_bases,
-		const std::vector<basis::ElementBases> &bases,
+		std::vector<basis::ElementBases> &bases,
```
Comment (teseoch) [inline src/polyfem/solver/SolveData.cpp:37]: really?

File: src/polyfem/solver/SolveData.cpp [hunk 3]
```diff
@@ -152,7 +154,7 @@ namespace polyfem::solver
 			if (damping_assembler != nullptr)
 			{
 				damping_form = std::make_shared<ElasticForm>(
-					n_bases, bases, geom_bases, *damping_assembler, ass_vals_cache, t, dt, is_volume);
+					n_bases, bases, geom_bases, *damping_assembler, ass_vals_cache, t, dt, is_volume, 0., "Discrete");
 				forms.push_back(damping_form);
```
Comment (teseoch) [inline src/polyfem/solver/SolveData.cpp:158]: string?

File: src/polyfem/solver/SolveData.hpp [hunk 4]
```diff
@@ -60,11 +62,13 @@ namespace polyfem::solver
 
 			// Elastic form
 			const int n_bases,
-			const std::vector<basis::ElementBases> &bases,
+			std::vector<basis::ElementBases> &bases,
```
Comment (teseoch) [inline src/polyfem/solver/SolveData.hpp:65]: are you sure?

File: src/polyfem/solver/forms/BCLagrangianForm.cpp [hunk 5]
```diff
@@ -51,6 +51,27 @@ namespace polyfem::solver
 			is_boundary_dof[bn] = false;
 
 		masked_lumped_mass_sqrt_ = mass.size() == 0 ? polyfem::utils::sparse_identity(ndof, ndof) : polyfem::utils::lump_matrix(mass);
+		{
+			double min_diag = std::numeric_limits<double>::max();
```
Comment (teseoch) [inline src/polyfem/solver/forms/BCLagrangianForm.cpp:55]: what is this?

File: src/polyfem/solver/forms/BCPenaltyForm.cpp [hunk 6]
```diff
@@ -52,6 +52,26 @@ namespace polyfem::solver
 
 		masked_lumped_mass_ = mass.size() == 0 ? polyfem::utils::sparse_identity(ndof, ndof) : polyfem::utils::lump_matrix(mass);
 		assert(ndof == masked_lumped_mass_.rows() && ndof == masked_lumped_mass_.cols());
+		{
```
Comment (teseoch) [inline src/polyfem/solver/forms/BCPenaltyForm.cpp:55]: what is this?

File: src/polyfem/solver/SolveData.cpp (hunk 3 above)
Comment (teseoch) [reply src/polyfem/solver/SolveData.cpp:158]: the 0 shouldbe from the json?

File: src/polyfem/solver/forms/ElasticForm.hpp [hunk 7]
```diff
@@ -96,5 +120,10 @@ namespace polyfem::solver
 		void compute_cached_stiffness();
 
 		Eigen::VectorXd x_prev_;
+
+		mutable std::vector<utils::Tree> quadrature_hierarchy_;
```
Comment (teseoch) [inline src/polyfem/solver/forms/ElasticForm.hpp:124]: mutable?

File: src/polyfem/state/StateSolveLinear.cpp [hunk 8]
```diff
@@ -174,7 +174,7 @@ namespace polyfem
 			n_bases, bases, geom_bases(),
 			*assembler, ass_vals_cache,
 			t, problem->is_time_dependent() ? args["time"]["dt"].get<double>() : 0.0,
-			mesh->is_volume());
+			mesh->is_volume(), 0, "Discrete");
```
Comment (teseoch) [inline src/polyfem/state/StateSolveLinear.cpp:177]: zero and string? should be enum? and maybe use defaults?

File: src/polyfem/utils/Jacobian.cpp [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+    {
+        const int order = std::max(bases[0].bases.front().order(), gbases[0].bases.front().order());
+        const int n_basis_per_cell = std::max(bases[0].bases.size(), gbases[0].bases.size());
+        Eigen::MatrixXd cp = extract_nodes(dim, bases, gbases, u, order);
+
+        bool flag = false;
+        unsigned invalid_id = 0;
+        Tree tree;
+
+        #define CHECK_STATIC(n,s,p) \
```
Comment (teseoch) [inline src/polyfem/utils/Jacobian.cpp:158]: why macros?

File: src/polyfem/utils/Jacobian.cpp [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+            for (int j = 0; j < mat.cols(); j++)
+            {
+                std::cout << mat(i, j);
+                if (i < mat.rows() - 1 || j < mat.cols() - 1)
+                    std::cout << ", ";
+            }
+        }
+    }
+
+    // void export_static_hdf5(
```
Comment (teseoch) [inline src/polyfem/utils/Jacobian.cpp:262]: delete?

File: src/polyfem/utils/Jacobian.cpp [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+
+        // logger().debug("Jacobian check order {}, number of nodes per cell {}, number of total nodes {}", order, n_basis_per_cell, cp2.rows());
+
+        unsigned invalid_id = -1;
+        bool gaveUp = false;
+        double step = 1;
+        double invalid_step = 1.;
+        Tree tree;
+
+        #define MAX_TIME_STEP(n,s,p) \
```
Comment (teseoch) [inline src/polyfem/utils/Jacobian.cpp:356]: macros?

File: src/polyfem/utils/Rational.hpp [hunk 12]
```diff
@@ -0,0 +1,280 @@
+#pragma once
```
Comment (teseoch) [inline src/polyfem/utils/Rational.hpp:1]: where is it used?

File: tests/test_form_derivatives.cpp [hunk 13]
```diff
@@ -317,7 +317,9 @@ TEST_CASE("elastic form derivatives", "[form][form_derivatives][elastic_form]")
 		state_ptr->ass_vals_cache,
 		0,
 		state_ptr->args["time"]["dt"],
-		state_ptr->mesh->is_volume());
+		state_ptr->mesh->is_volume(),
+		0.,
```
Comment (teseoch) [inline tests/test_form_derivatives.cpp:321]: revert all of these with defaults

**Review by teseoch — COMMENTED** [review]
we need integration tests

File: src/polyfem/solver/forms/ElasticForm.hpp (hunk 7 above)
Comment (Huangzizhou) [reply src/polyfem/solver/forms/ElasticForm.hpp:124]: Adaptive refinement happens inside the step valid check, which is a const call

File: src/polyfem/assembler/MultiModel.hpp [hunk 14]
```diff
@@ -39,6 +39,7 @@ namespace polyfem::assembler
 		inline void init_multimodels(const std::vector<std::string> &mats) { multi_material_models_ = mats; }
 
 		std::string name() const override { return "MultiModels"; }
+		bool allow_inversion() const override { return true; }
```
Comment (teseoch) [inline src/polyfem/assembler/MultiModel.hpp:42]: this should check all the models?

**Review by teseoch — COMMENTED** [review]
Thanks! In the output data many new lines are not tested, can you enable them in some integration tests?

**Pull request merged**